File: Solutions/40. Combination Sum II/40. Combination Sum II.cpp
class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& nums, int target) {
        vector<vector<int>> res;
        vector<int> tmp;
        sort(nums.begin(),nums.end());
        dfs(res,tmp,nums,target,0);
        return res;
    }
    void dfs(vector<vector<int>>& res,vector<int>& tmp,vector<int>& nums,int target,int st){
        if(target<0){
            return;
        }
        if(target==0){
            res.push_back(tmp);
            return;
        }
        for(int i=st;i<nums.size();++i){
            if(i>st and nums[i]==nums[i-1]){
                continue;
            }
            tmp.push_back(nums[i]);
            dfs(res,tmp,nums,target-nums[i],i+1);
            tmp.pop_back();
        }
    }
};

File: Solutions/40. Combination Sum II/40. Combination Sum II.go
func combinationSum2(candidates []int, target int) [][]int {
    var res [][]int
    var path []int
    sort.Ints(candidates)
    foo(candidates,target,&res,path,0)
    return res
}
func foo(c []int,t int,res *[][]int,path []int,begin int){
    if t<=0{
        if t==0{
            tmp:=make([]int,len(path))
            copy(tmp,path)
            *res=append(*res,tmp)
            return
        }
        return
    }
    for i:=begin;i<len(c);i+=1{
        if i>begin && c[i]==c[i-1]{
            continue
        }
        path=append(path,c[i])
        foo(c,t-c[i],res,path,i+1)
        path=path[:len(path)-1]
    }
}